#1162 add openApi documentation to beacon state handler

rolfyone opened the pull request:
Add an object level abstraction for the response object, so that openApi can render what the response will be similar to.

 - The openApi was not able to handle SZVector or SZList.
 - It also defined Byte32 in a messy way, and our representation of Byte32 is actually String
 - UnsignedLong was coming out as an empty object, not any kind of numeric representation.

 The one messy thing about these objects is they're all 'public', because private members don't get rendered in openApi.  This means we have objects that are 'similar' and just used for documentation, which I'm not the hugest fan of, but it does look like we don't have a lot of choices.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
#1162 add openApi documentation to beacon state handler

Add an object level abstraction for the response object, so that openApi can render what the response will be similar to.

 - The openApi was not able to handle SZVector or SZList.
 - It also defined Byte32 in a messy way, and our representation of Byte32 is actually String
 - UnsignedLong was coming out as an empty object, not any kind of numeric representation.

 The one messy thing about these objects is they're all 'public', because private members don't get rendered in openApi.  This means we have objects that are 'similar' and just used for documentation, which I'm not the hugest fan of, but it does look like we don't have a lot of choices.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
fixes from looking at draft review.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
fixes from looking at draft review.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconStateHandler.java:
```diff
@@ -13,26 +13,62 @@
+    String rootParam = ctx.queryParam("root");
+    BeaconState result = queryByRootHash(rootParam);
+    if (result == null) {
+      LOG.debug("Block root {} not found", rootParam);
```
Suggest trace level not debug because it's essentially just recording the flow control that was followed.

rolfyone replied:
ok

ajsutton commented on data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconStateHandler.java:
```diff
@@ -13,26 +13,62 @@
+      responses = {
+        @OpenApiResponse(
+            status = "200",
+            content = @OpenApiContent(from = BeaconStateResponse.class)),
```
As discussed separately, I think we want to avoid having the classes used to define the API documentation being separate from the classes used to actually create the responses to avoid them getting out of sync.  If we do need to separate them we may as well just generate the OpenAPI documentation directly and not have to deal with the complication of defining OpenAPI semantics in Java code and annotations.

ajsutton commented on data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconStateHandler.java:
```diff
@@ -13,26 +13,62 @@
-    return "/beacon/state";
+  private BeaconState queryByRootHash(String root) {
+    Bytes32 root32 = Bytes32.fromHexString(root);
+    return client.getStore().getBlockState(root32);
```
Pre-genesis `client.getStore()` will return `null` so we need to handle that case.  I suspect you just want to use `client.getBlockState`.

rolfyone replied:
ok, i think i've captured that next commit.

ajsutton commented on data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconStateHandlerTest.java:
```diff
@@ -0,0 +1,62 @@
+public class BeaconStateHandlerTest {
+
+  private final ChainStorageClient mockStorageClient = mock(ChainStorageClient.class);
+  private final Store mockStore = mock(Store.class);
```
nit: We don't generally use a `mock` prefix for mocks in the Teku codebase.

rolfyone replied:
nit: you probably should! it can get super confusing when some are mocked and some aren't! 

I'll resolve the ones i know about in my branch, as part of this commit.

rolfyone committed:
WIP changed back to Bytes32 and UnsignedLong after discussion.

Started to add constructors with a view of potentially using these objects as the actual interface results.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
address early review comments.

Signed-off-by: Paul Harris <paul.harris@consensys.net>
